### PR TITLE
fix: make fields sorted by lexicographical order

### PR DESCRIPTION
**Pull request by dateri**
We fix that some fields are in random order (issue #7710). We use `Treemap` which is ordered to replace `LinkedHashMap` and added a comparison function to the treemap to ensure that fields are ordered by lexicographical order.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Normally, this would be a suitable way. However, this is a special case as you can reorder the fields in the custom entry types dialog. Refs #6152
The TreeSet will now kill any individual order. What we only need is a sorting in the dropdown box. So you need to add a sorting in the GUI/ViewModel

**Review by Siedlerchr — CHANGES_REQUESTED** [review]
Do not change the order in FieldFactory
Edit:: Please change the PR title to a more meaningful name!

**Comment by Siedlerchr** [comment]
Normally, this would be a suitable way. However, this is a special case as you can reorder the fields in the custom entry types dialog. Refs #6152
The TreeSet will now kill any individual order. What we only need is a sorting in the dropdown box. So you need to add a sorting in the GUI/ViewModel

**Comment by dateri** [comment]
Thanks for guidance, I found they both use `src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java`, so I change it in the cleanupsPanel.

File: src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java [hunk 1]
```diff
@@ -28,6 +28,12 @@
     private final ObjectProperty<Formatter> selectedFormatterProperty = new SimpleObjectProperty<>();
 
     public FieldFormatterCleanupsPanelViewModel() {
+        availableFieldsProperty.sort((field1, field2) -> {
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:31]: Can you use `Comparator.comparing` instead?

File: CHANGELOG.md [hunk 2]
```diff
@@ -550,6 +550,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where the XMP Importer would incorrectly return an empty default entry when importing pdfs [#6577](https://github.com/JabRef/jabref/issues/6577)
 - We fixed an issue where opening the menu 'Library properties' marked the library as modified [#6451](https://github.com/JabRef/jabref/issues/6451)
 - We fixed an issue when importing resulted in an exception [#7343](https://github.com/JabRef/jabref/issues/7343)
+- We fixed an issue Some fields are in random order [#7710](https://github.com/JabRef/jabref/issues/7710)
```
Comment (Siedlerchr) [inline CHANGELOG.md:553]:
```suggestion
- We fixed an issue where the field in the Field formatter dropdown selection were sorted in random order [#7710](https://github.com/JabRef/jabref/issues/7710)
```

File: src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java [hunk 3]
```diff
@@ -28,6 +28,12 @@
...
+            if (field1.getDisplayName().equals(field2.getDisplayName())) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:32]: The compareTo already performs equal checks...
So you should be able to directly use sort(field1, field2) -> field1.getDisplayName().compareTo(field2.getDisplayName))

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:32]: `Comparator.comparing` should work as well.

Comment (dateri) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: What you said really makes sense and I have modified them.

File: src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java [hunk 4]
```diff
@@ -17,6 +17,8 @@
 import org.jabref.model.entry.field.Field;
 import org.jabref.model.entry.field.FieldFactory;
 
+import java.util.Comparator;
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:20]: Please fix the reviewdog complaint

File: src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java [hunk 5]
```diff
@@ -28,6 +30,7 @@
     private final ObjectProperty<Formatter> selectedFormatterProperty = new SimpleObjectProperty<>();
 
     public FieldFormatterCleanupsPanelViewModel() {
+        availableFieldsProperty.sort(Comparator.comparing(Field::getDisplayName));
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:33]: Nitpick: In this case, it doesn't change anything, but you can use a [SortedList](https://openjfx.io/javadoc/16/javafx.base/javafx/collections/transformation/SortedList.html#%3Cinit%3E(javafx.collections.ObservableList,java.util.Comparator)) to wrap the `ObservableList` in instead. If at some time in the future the underlying data structure (`FieldFactory.getCommonFields()`) becomes observable this menu will be kept sorted.

**Review by k3KAW8Pnf7mkmdSMPHz27 — CHANGES_REQUESTED** [review]
Looks good to me! One needed change and one nitpick.

(nitpick = I will approve even if you don't fix it, it is a suggestion and address it if you like it)

File: src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java (hunk 5 above)
Comment (dateri) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:33]: Emmm... you mean make `availableFieldsProperty` as `SortedList` ?

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java:33]: Oups, I might have commented on the wrong line

Comment (dateri) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: It seems that the Comparator of the sortedList must be chosen at definition? `UnsupportedOperationException` will be throw n if I sort without Comparator. So i wonder whether there is no need to sort in `FieldFormatterCleanupsPanelViewModel`.

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: I am not sure that I understand you. I believe [setComparator](https://openjfx.io/javadoc/16/javafx.base/javafx/collections/transformation/SortedList.html#setComparator(java.util.Comparator)) should work, but why do you want to set it in the constructor?

Comment (dateri) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: I mean should we still use `availableFieldsProperty.sort(Comparator.comparing(Field::getDisplayName));` after using a sortList ?

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: I can't find a good reference for it, but no, that should not be needed. `SortedList` should keep the list sorted based on the `Comparator` as long as you are making changes that fire change events (e.g., add, remove).

Comment (dateri) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: OK. Very thanks for your help!

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/gui/commonfxcontrols/FieldFormatterCleanupsPanelViewModel.java]: [This](https://stackoverflow.com/questions/27806814/javafx-sortedlist-listen-the-list-changes-and-the-list-items-updated-event#tab-top) is an example of when it must be dealt with differently.

And thank you for including the nitpick 👍

**Review by k3KAW8Pnf7mkmdSMPHz27 — APPROVED** [review]
LTGM!

**Review by Siedlerchr — COMMENTED** [review]
Yeah, that is even a better solution!

**Review by Siedlerchr — APPROVED** [review]

**Pull request merged**